homeassistant/components/telegram_bot/__init__.py

```diff
@@ -373,49 +402,91 @@
 PLATFORMS: list[Platform] = [Platform.EVENT, Platform.NOTIFY]
 
 
-async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:  # noqa: C901 ignore complexity caused by _call_service()
     """Set up the Telegram bot component."""
 
     async def async_send_telegram_message(service: ServiceCall) -> ServiceResponse:
         """Handle sending Telegram Bot message service calls."""
 
-        msgtype = service.service
-        kwargs = dict(service.data)
-        _LOGGER.debug("New telegram message %s: %s", msgtype, kwargs)
-
-        if ATTR_TIMEOUT in service.data:
-            _deprecate_timeout(hass, service)
-
-        config_entry_id: str | None = service.data.get(CONF_CONFIG_ENTRY_ID)
-        config_entry: TelegramBotConfigEntry | None = None
-        if config_entry_id:
-            config_entry = hass.config_entries.async_get_known_entry(config_entry_id)
+        _deprecate_timeout(hass, service)
 
-        else:
-            config_entries: list[TelegramBotConfigEntry] = (
-                service.hass.config_entries.async_entries(DOMAIN)
-            )
+        # this is the list of targets to send the message to
+        targets: list[tuple[TelegramBotConfigEntry, int, str | None]] = _build_targets(
+            hass, service
+        )
 
-            if len(config_entries) == 1:
-                config_entry = config_entries[0]
+        service_responses: JsonValueType = []
+        errors: list[tuple[HomeAssistantError, str]] = []
 
-            if len(config_entries) > 1:
+        # invoke the service for each target
+        for target_config_entry, target_chat_id, target_notify_entity_id in targets:
+            if not hasattr(target_config_entry, "runtime_data"):
                 raise ServiceValidationError(
-                    "Multiple config entries found. Please specify the Telegram bot to use.",
+                    "No config entries found or setup failed. Please set up the Telegram Bot first.",
                     translation_domain=DOMAIN,
-                    translation_key="multiple_config_entry",
+                    translation_key="missing_config_entry",
                 )
 
-        if not config_entry or not hasattr(config_entry, "runtime_data"):
-            raise ServiceValidationError(
-                "No config entries found or setup failed. Please set up the Telegram Bot first.",
+            try:
+                service_response = await _call_service(
+                    service, target_config_entry.runtime_data, target_chat_id
+                )
+
+                if service.service == SERVICE_DOWNLOAD_FILE:
+                    return service_response
+
+                if service_response is not None:
+                    formatted_response: list[JsonValueType] = [
+                        {
+                            ATTR_CHAT_ID: int(chat_id),
+                            ATTR_MESSAGEID: message_id,
```

Comment:
Side note: I'm always annoyed when I read `ATTR_MESSAGEID`. Would be nice to separate the words message and id with an underscore in a different PR.